### PR TITLE
AKCORE-144: Implemented share-consumer-fetch-manager-metrics

### PR DESCRIPTION
**Pull request by ShivsundarR**
*What*
Implemented share-consumer-fetch-manager-metrics listed in the KIP.
The Acknowledgement metrics are yet to be recorded in this PR. They will be implemented after the revised ShareFetchRequestManager is pushed.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchMetricsManager.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    public final MetricName recordsFetchedRate;
+    public final MetricName recordsFetchedTotal;
+    public final MetricName recordsPerRequestAvg;
+    public final MetricName recordsPerRequestMax;
+
+    public ShareFetchMetricsManager(Metrics metrics, FetchMetricsRegistry fetchMetricsRegistry) {
+        this.metricsRegistry = fetchMetricsRegistry;
+
+        String grpName = SHARE_CONSUMER_METRIC_GROUP_PREFIX + "-fetch-manager-metrics";
+
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchMetricsManager.java:50]: In comparison with the regular fetch metrics, the share-consumer-fetch-manager-metrics are all "client level". We might add topic level and partition level in a future KIP, so I would at least add a single-line comment `/* Client level */` to make the correspondence with FetchMetricsRegistry more clear.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchMetricsManager.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    public final MetricName acknowledgementSendRate;
+    public final MetricName acknowledgementSendTotal;
+    public final MetricName acknowledgementErrorRate;
+    public final MetricName acknowledgementErrorTotal;
+    public final MetricName recordsFetchedRate;
+    public final MetricName recordsFetchedTotal;
+    public final MetricName recordsPerRequestAvg;
+    public final MetricName recordsPerRequestMax;
+
+    public ShareFetchMetricsManager(Metrics metrics, FetchMetricsRegistry fetchMetricsRegistry) {
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchMetricsManager.java:46]: Why did you not make a ShareFetchMetricsRegistry?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchMetricsManager.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        this.metricsRegistry = fetchMetricsRegistry;
+
+        String grpName = SHARE_CONSUMER_METRIC_GROUP_PREFIX + "-fetch-manager-metrics";
+
+        acknowledgementSendRate = metrics.metricName("acknowledgements-send-rate", grpName,
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchMetricsManager.java:51]: I would prefer to see these in the same order as the KIP for ease of review.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchMetricsManager.java (hunk 2 above)
Comment (ShivsundarR) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchMetricsManager.java]: I had initially thought so as lot of metrics were already present in the existing FetchMetricsRegistry. But yeah, as we had discussed on-call, this is a small piece of code and the design would be cleaner with more metrics in future if we have a separate ShareFetchMetricsRegistry. Have made the changes.

**Review by AndrewJSchofield — APPROVED** [review]

**Pull request merged**